File: src/levels/L21-null-object/null_object.rb
class LoyaltyCards
  # region: LoyaltyCards#find
  def find(customer) = @cards.fetch(customer.id) { NullCard.new }
  # endregion
end

# Responde a todo lo que responde una tarjeta, sin hacer nada. Nadie necesita preguntar "¿es nil?".
# region: NullCard
class NullCard
  def add_points(_amount) = nil
  def discount = 0
  def masked_number = "sin tarjeta"
end
# endregion

File: src/levels/L21-null-object/ts/null_object.ts
class LoyaltyCards {
  private cards = new Map<string, Card>()

  // region: LoyaltyCards#find
  find(customer: Customer): Card { return this.cards.get(customer.id) ?? new NullCard() }
  // endregion
}

// Responde a todo lo que responde una tarjeta, sin hacer nada. Nadie necesita preguntar "¿es undefined?".
// region: NullCard
class NullCard implements Card {
  addPoints(_amount: number) {}
  get discount() { return 0 }
  get maskedNumber() { return 'sin tarjeta' }
}
// endregion
